langpuppet/gapi.go

```diff
@@ -0,0 +1,284 @@
+
+	var childFlags []cli.Flag
+
+	for _, flag := range append(langFlags, puppetFlags...) {
+		childFlags = append(childFlags, &cli.StringFlag{
+			Name:  FlagPrefix + strings.Split(flag.GetName(), ",")[0],
+			Value: "",
+			Usage: fmt.Sprintf("equivalent for '%s' when using the lang/puppet entrypoint", flag.GetName()),
+		})
+	}
+
+	return childFlags
+}
+
+// Init initializes the langpuppet GAPI struct.
+func (obj *GAPI) Init(data gapi.Data) error {
+	if obj.initialized {
+		return fmt.Errorf("already initialized")
+	}
+	obj.data = data // store for later
+
+	dataLang := gapi.Data{
+		Program:       obj.data.Program,
+		Hostname:      obj.data.Hostname,
+		World:         obj.data.World,
+		Noop:          obj.data.Noop,
+		NoConfigWatch: obj.data.NoConfigWatch,
+		NoStreamWatch: obj.data.NoStreamWatch,
+		Debug:         obj.data.Debug,
+		Logf: func(format string, v ...interface{}) {
+			obj.data.Logf(lang.Name+": "+format, v...)
+		},
+	}
+
+	dataPuppet := gapi.Data{
+		Program:       obj.data.Program,
+		Hostname:      obj.data.Hostname,
+		World:         obj.data.World,
+		Noop:          obj.data.Noop,
+		NoConfigWatch: obj.data.NoConfigWatch,
+		NoStreamWatch: obj.data.NoStreamWatch,
+		Debug:         obj.data.Debug,
+		Logf: func(format string, v ...interface{}) {
+			obj.data.Logf(puppet.Name+": "+format, v...)
+		},
+	}
+
+	if err := obj.langGAPI.Init(dataLang); err != nil {
+		return err
+	}
+	if err := obj.puppetGAPI.Init(dataPuppet); err != nil {
+		return err
+	}
+
+	obj.closeChan = make(chan struct{})
+	obj.initialized = true
+	return nil
+}
+
+// Graph returns a current Graph.
+func (obj *GAPI) Graph() (*pgraph.Graph, error) {
+	if !obj.initialized {
+		return nil, fmt.Errorf("%s: GAPI is not initialized", Name)
+	}
+
+	var graph_err error
+
+	if obj.currentLangGraph == nil {
+		obj.currentLangGraph, graph_err = obj.langGAPI.Graph()
+		if graph_err != nil {
+			return nil, graph_err
+		}
+	}
+	if obj.currentPuppetGraph == nil {
+		obj.currentPuppetGraph, graph_err = obj.puppetGAPI.Graph()
+		if graph_err != nil {
+			return nil, graph_err
+		}
+	}
+
+	g, err := mergeGraphs(obj.currentLangGraph, obj.currentPuppetGraph)
+
+	obj.data.Logf("current graph from lang:")
+	obj.currentLangGraph.Logf(obj.data.Logf)
+	obj.data.Logf("current graph from puppet:")
+	obj.currentPuppetGraph.Logf(obj.data.Logf)
+	if err == nil {
+		obj.data.Logf("merge result:")
+		g.Logf(obj.data.Logf)
+	}
+
+	return g, err
+}
+
+// Next returns nil errors every time there could be a new graph.
+func (obj *GAPI) Next() chan gapi.Next {
+	ch := make(chan gapi.Next)
+	obj.wg.Add(1)
+	go func() {
+		defer obj.wg.Done()
+		defer close(ch) // this will run before the obj.wg.Done()
+		if !obj.initialized {
+			next := gapi.Next{
+				Err:  fmt.Errorf("%s: GAPI is not initialized", Name),
+				Exit: true, // exit, b/c programming error?
+			}
+			ch <- next
+			return
+		}
+		nextLang := obj.langGAPI.Next()
+		nextPuppet := obj.puppetGAPI.Next()
+
+		langReady := false
```

Comment:
The logic in Next() looks perfect... Except I would expect these "ready" flags to be used in the Graph method to decide which Graph sub method to call... I'm not sure if you'd need a lock or not. I'd need to think about it. (Safer, but can we deadlock??)

Reply:
Raciness is an interesting aspect, but I don't see the benefit of carrying another flag, versus "pointer is now nil" bearing the same semantics as a "dirty" flag.

Reply:
It's *very* different, because *after* the _first_ run of both, then both of those "flags" are not nil (they contain graphs) and they NEVER go nil again. So a "dirty" operation never actually does anything. Get it?

As a result for langGAPI to return on Next() and you call graph on both, then you've just panic or blocked the system because Puppet wasn't ready for you to call Graph. Get it?

Reply:
Doing it with the ready flag now.